File: Python/LinkedList/Intersection in Y Shaped Lists.py
//https://www.geeksforgeeks.org/problems/intersection-point-in-y-shapped-linked-lists/1
def getLength(root):
    cnt = 0
    temp = root
    while temp != None:
        temp = temp.next
        cnt += 1
    
    return cnt

def intersetPoint(head1,head2):
    l1, l2 = getLength(head1), getLength(head2)
    
    if l1 < l2:
        head1, head2 = head2, head1
        l1, l2 = l2, l1
    
    diff = l1 - l2
    
    ptr1, ptr2 = head1, head2
    
    while diff > 0:
        ptr1 = ptr1.next
        diff -= 1
    
    while ptr1 != None:
        if ptr1 == ptr2:
            return ptr1.data
        
        ptr1 = ptr1.next
        ptr2 = ptr2.next
    
    return -1

File: Java/Linked List/Intersection in Y Shaped Lists.java
//https://www.geeksforgeeks.org/problems/intersection-point-in-y-shapped-linked-lists/1
class Intersect {
    // Function to find intersection point in Y shaped Linked Lists.
    static Node intersectPoint(Node head1, Node head2) {
        int count1 = 0, count2 = 0;
        Node temp1 = head1, temp2 = head2;
        
        // Find the length of the first linked list
        while (temp1 != null) {
            count1++;
            temp1 = temp1.next;
        }
        
        // Find the length of the second linked list
        while (temp2 != null) {
            count2++;
            temp2 = temp2.next;
        }
        
        // Determine the difference in lengths
        int diff = Math.abs(count1 - count2);
        
        // Reset the pointers to the heads of the lists
        temp1 = head1;
        temp2 = head2;
        
        // Move the pointer of the longer list ahead by the difference in lengths
        if (count1 > count2) {
            while (diff-- > 0) {
                temp1 = temp1.next;
            }
        } else {
            while (diff-- > 0) {
                temp2 = temp2.next;
            }
        }
        
        // Traverse both lists simultaneously and check for intersection
        while (temp1 != null && temp2 != null) {
            if (temp1 == temp2) {
                return temp1;  // Return the intersection node
            }
            temp1 = temp1.next;
            temp2 = temp2.next;
        }
        
        return null;  // No intersection found
    }
}
